### ISSUE TITLE
no-matching-overload on pd.read_csv(data, usecols=['foo', 'bar'])

### ISSUE THREAD

**Issue by darabos** [astral-sh/ty#2590]
### Summary

Example code:

```python
import pandas as pd

pd.read_csv("data.csv", usecols=["foo", "bar"])
```

This is exactly what the [Pandas docs](https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html) tell me to use. With both Pandas 2.2.3 and Pandas 3.0.0 ty gives me:

```
error[no-matching-overload]: No overload of function `read_csv` matches arguments
 --> x.py:3:1
  |
1 | import pandas as pd
2 |
3 | pd.read_csv("data.csv", usecols=["foo", "bar"])
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
info: First overload defined here
   --> .venv/lib/python3.12/site-packages/pandas/io/parsers/readers.py:631:5
    |
629 |   # iterator=True -> TextFileReader
630 |   @overload
631 |   def read_csv(
    |  _____^
[...snip...]
```

This code passed the check in ty 0.0.1a21 but is failing in 0.0.1a22 and later, including ty 0.0.13.

It looks like the type for `usecols` is pretty complicated:
https://github.com/pandas-dev/pandas/blob/v3.0.0/pandas/_typing.py#L530

The matching type from that union is supposed to be the SequenceNotStr protocol. So I think this is also an example of #1714.

**Comment by AlexWaygood** [comment]
Thanks! I initially struggled to reproduce this, as it does not reproduce if you have any version of `pandas-stubs` installed or `pandas>=3` installed:

```
~/dev % uvx --with=pandas ty check foo.py
Installed 5 packages in 25ms
All checks passed!
~/dev % uvx --with=pandas-stubs ty check foo.py
Installed 4 packages in 14ms
All checks passed!
~/dev % uvx --with=pandas-stubs ty@0.0.12 check foo.py
Installed 4 packages in 9ms
All checks passed!
~/dev % uvx --with=pandas-stubs ty@0.0.11 check foo.py
Installed 4 packages in 9ms
All checks passed!
~/dev % uvx --with=pandas-stubs ty@0.0.10 check foo.py
Installed 4 packages in 43ms
All checks passed!
~/dev % uvx --with=pandas ty@0.0.10 check foo.py
Installed 5 packages in 16ms
All checks passed!
~/dev % uvx --with=pandas ty@0.0.11 check foo.py
Installed 5 packages in 16ms
All checks passed!
~/dev % uvx --with=pandas ty@0.0.12 check foo.py
Installed 5 packages in 16ms
All checks passed!
~/dev % uvx --with=pandas ty@0.0.13 check foo.py
All checks passed!
~/dev % cat foo.py                                                                  
import pandas as pd

pd.read_csv("data.csv", usecols=["foo", "bar"])
```

From looking at your repo, however, I see that you have `pandas==2.2.3` pinned in your `uv.lock` file. I am indeed able to reproduce this diagnostic if I install `pandas<3` into an environment that does not also have `pandas-stubs` installed.

However, all other type checkers appear to emit errors on that file as well if you only have `pandas==2.2.3` installed and not `pandas-stubs`, which makes me question whether this is a bug in ty:

```
~/dev [1] % uvx --with='pandas==2.2.3' pyright foo.py      
Installed 9 packages in 105ms
/Users/alexw/dev/foo.py
  /Users/alexw/dev/foo.py:3:1 - error: No overloads for "read_csv" match the provided arguments (reportCallIssue)
  /Users/alexw/dev/foo.py:3:33 - error: Argument of type "list[str]" cannot be assigned to parameter "usecols" of type "UsecolsArgType[Unknown]" in function "read_csv"
    Type "list[str]" is not assignable to type "UsecolsArgType[Unknown]"
      "list[str]" is incompatible with protocol "SequenceNotStr[Hashable]"
        "index" is an incompatible type
          Type "(value: str, start: SupportsIndex = 0, stop: SupportsIndex = sys.maxsize, /) -> int" is not assignable to type "(value: Any, /, start: int = 0, stop: int = ...) -> int"
            Missing keyword parameter "start"
            Missing keyword parameter "stop"
      "list[str]" is not assignable to "range"
      "list[str]" is not assignable to "ExtensionArray"
    ... (reportArgumentType)
2 errors, 0 warnings, 0 informations
~/dev [1] % uvx --with='pandas==2.2.3' mypy foo.py   
Installed 11 packages in 80ms
foo.py:3: error: Module has no attribute "read_csv"  [attr-defined]
Found 1 error in 1 file (checked 1 source file)
~/dev [1] % uvx --with='pandas==2.2.3' pyrefly check foo.py
Installed 7 packages in 61ms
 WARN Failed to query interpreter at /Users/alexw/dev/pandas-stubs/.venv/bin/python3, falling back to default Python environment settings
No such file or directory (os error 2)
 WARN Failed to query interpreter at /Users/alexw/dev/pandas-stubs/.venv/bin/python3, falling back to default Python environment settings
No such file or directory (os error 2)
ERROR No attribute `read_csv` in module `pandas` [missing-attribute]
 --> foo.py:3:1
  |
3 | pd.read_csv("data.csv", usecols=["foo", "bar"])
  | ^^^^^^^^^^^
  |
 INFO 1 error
```

In general, I would advise installing `pandas-stubs` when type-checking code that uses pandas. If I understand correctly, the annotations inside the pandas repository are still incomplete and mainly intended for internal CI use rather than for external users of the pandas library.

**Comment by AlexWaygood** [comment]
As the message from pyright indicates, the problem is that `list[str]` is not compatible with the `SequenceNotStr` protocol as it existed in v2.2.3. If you look at the pandas source code on that tag, `SequenceNotStr` states that a concrete class can only be a subtype of that protocol if the concrete class's `index` method has `start` and `stop` parameters that accept keyword arguments. `list.index` does not satisfy that interface, so it does not satisfy the v2.2.3 version of `SequenceNotStr`, so it is not a subtype of the `UsecolsArgType` type alias, so no overloads match.

ty is working correctly here; it's pandas's type annotations that were broken. The solution is to install pandas-stubs when running ty or to upgrade your pinned version of pandas.

**Comment by darabos** [comment]
Thank you! Sorry, I thought I checked with Pandas 3.0.0, but I didn't. (`uv run` switched me back to 2.2.3.)
